site/topics/programming/assembler.py

```diff
@@ -0,0 +1,79 @@
+import sys
+
+OPERATORS = {
+    "NOOP": 0b0000,
+    "LDAR": 0b0001,
+    "LDAD": 0b0010,
+    "LDBR": 0b0011,
+    "LDBD": 0b0100,
+    "SAVA": 0b0101,
+    "SAVB": 0b0110,
+    "ADDB": 0b0111,
+    "SUBB": 0b1000,
+    "JMPA": 0b1001,
+    "OUTU": 0b1101,
+    "OUTS": 0b1110,
+    "HALT": 0b1111,
+}
+HAS_OPERAND = {
+    "LDAR",
+    "LDAD",
+    "LDBR",
+    "LDBD",
+    "SAVA",
+    "SAVB",
+    "JMPA",
+    "OUTU",
+    "OUTS",
+}
+
+def parse_number(number_string:str) -> int:
+    try:
+        if number_string[:2] == "0b":
+            number = int(number_string, 2)
+        elif number_string[:2] == "0x":
+            number = int(number_string, 16)
+        else:
+            number = int(number_string)
+    except ValueError:
+        raise ValueError(f"Cannot parse operand {number_string}")
+    return number
+
+
+file_to_assemble = sys.argv[1]
+with open(file_to_assemble) as file:
+    program_list = file.read().strip().split("\n")
+
+if len(program_list) > 16:
+    raise ValueError(f"Program length of {len(program_list)} exceeds maximum size of 16 bytes")
+
+machine_code = [0x00] * 16
+for i, raw_program_line in enumerate(program_list):
+    line = raw_program_line.split()
+    if line[0].isalpha():
+        try:
+            operator = OPERATORS[line[0]]
+        except KeyError:
+            raise ValueError(f"Unknown operator {line[0]}")
+        if line[0] not in HAS_OPERAND:
+            if len(line) > 1:
+                raise ValueError(f"Operator {line[0]} requires no operand")
```

Comment:
You won't hear me say this much, but it's probably simpler to use a regex.
```python
syntax = {
  r"NOOP",
  r"LDAR\s+\b(0x[0-9a-fA-F]|0b[0-1]+|[0-9]+)\b",
  # etc
}
```
Then `re.match(one_of_the_syntax_options, line.trim())` is either `None` if it failed, or a `Match` where `.group(0)` gives you have the entire line, like `LDAR    0xf`. Loop over all the syntax until you find a match, or crash out if no match is found. Then you can `line.split()` and do the rest of the logic.

Reply:
Although the hex number as an operand should never exceed 1 hex-digit, I allowed it to be more in the regex such that the exception will be thrown later when checking that the value does not exceed 16, that way the message is a little more specific. Besides, it should then check the binary is no more than 4 bits and then decimal would be weird. By allowing hex to be more than 1 hex-digit, it's consistent for the other number bases. 

`r"LDAR\s+\b(0x[0-9a-fA-F]+|0b[0-1]+|[0-9]+)\b"`